### PR TITLE
[fix] [broker] fix incorrect delete sub when lastActive do not update.

### PR DESCRIPTION
**Pull request by thetumbled**
### Motivation
 PR https://github.com/apache/pulsar/pull/17573 has fixed cases that consumer commit offset periodically. It update `lastActive` field of cursor when consumer commit the offset.
But there is still corner case that the consumer do not commit the offset for any reason. For example,  there is no content in the topic could be read, so the consumer do not commit the offset reasonably. In such case we has no reason delete the subscription.
Anyway, we should be cautious with the delete operation, which could result into data duplication or loss.

### Modifications

Update the `lastActive` field when check the inactive sub.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/thetumbled/pulsar/pull/31

### CODE REVIEW

**Review by poorbarcode — COMMENTED** [review]
@thetumbled Could you add a test for this case? Thanks

**Review by Technoboy- — CHANGES_REQUESTED** [review]
why do we update the cursor when every check ?

**Comment by thetumbled** [comment]
> why do we update the cursor when every check ?

If the consumer do not commit the offset for any reason, the `lastActive` will never be updated, which will result into incorrect subscription deletion.

**Comment by Technoboy-** [comment]
> > why do we update the cursor when every check ?
> 
> If the consumer do not commit the offset for any reason, the `lastActive` will never be updated, which will result into incorrect subscription deletion. As we know that the consumer is connected, we should update `lastActive` when doing check.

in this case, the consumer is connected, how could the code run into `sub.delete`
https://github.com/apache/pulsar/blob/8beac8b12ef7c0ef54529fbb7e4e76c54dea6283/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L2982-L2985

**Comment by thetumbled** [comment]
> > > why do we update the cursor when every check ?
> > 
> > 
> > If the consumer do not commit the offset for any reason, the `lastActive` will never be updated, which will result into incorrect subscription deletion. As we know that the consumer is connected, we should update `lastActive` when doing check.
> 
> in this case, the consumer is connected, how could the code run into `sub.delete`
> 
> https://github.com/apache/pulsar/blob/8beac8b12ef7c0ef54529fbb7e4e76c54dea6283/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L2982-L2985

If the broker1 shutdown ungracefully, the topic will be loaded on broker2 without update the `lastActive` field to zk, and the broker2 load  the topic with old `lastActive` field from zk. If broker2 check the inactive sub before any consumer connect to this sub, the sub will be deleted. 
This problem is simillar to https://github.com/apache/pulsar/pull/17573.

**Comment by thetumbled** [comment]
> @thetumbled Could you add a test for this case? Thanks

I push a new commit to add the test code, PTAL, thanks.

**Comment by thetumbled** [comment]
PTAL, thanks. @dao-jun @lhotari @Technoboy- @poorbarcode @coderzc

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceTest.java [hunk 1]
```diff
@@ -1313,6 +1314,66 @@ public void testCheckInactiveSubscriptionsShouldNotDeleteCompactionCursor() thro
...
+                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest).subscribe();
+        consumer.acknowledge(consumer.receive());
+
+        Optional<Topic> topicOptional = pulsar.getBrokerService().getTopic(topic, true).get();
+        assertTrue(topicOptional.isPresent());
+        PersistentTopic persistentTopic = (PersistentTopic) topicOptional.get();
+
+        // wait for 1min, but consumer is still connected all the time.
+        // so subscription should not be deleted.
+        Thread.sleep(60000);
```
[the remark this answers is not in the document]
Comment (thetumbled) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceTest.java:1345]: It is, but the minimum expiration time is 1min, we have to wait for 1min to trigger the bug.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceTest.java [hunk 2]
```diff
@@ -1313,6 +1314,66 @@ public void testCheckInactiveSubscriptionsShouldNotDeleteCompactionCursor() thro
...
+        persistentTopic.checkInactiveSubscriptions();
+        assertTrue(persistentTopic.getSubscriptions().containsKey("sub1"));
+        PersistentSubscription sub = persistentTopic.getSubscription("sub1");
+
+        // shutdown pulsar ungracefully
+        // disable the updateLastActive method to simulate the ungraceful shutdown
```
[the remark this answers is not in the document]
Comment (thetumbled) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceTest.java:1351]: We can't ensure that the updated timestamp is persisted in crash time. With this PR, we can fix most of the cases, as we update it every time broker do expiration check.
But strictly speaking, we can't fix all the cases. If we fail to persist the cursor info, the incorrect sub deletion occurs too.
Incorrect deletion of subscription causes serious consequence, such as duplication or data lost. In my opinion, missing deletions are better than incorrect deletion, this is PR https://github.com/apache/pulsar/pull/22794 try to do.

**Review by liangyepianzhou — CHANGES_REQUESTED** [review]
Update last active when a consumer disconnect instead of every check.

**Comment by thetumbled** [comment]
closed as https://github.com/apache/pulsar/pull/22794 has been merged.

**Pull request closed without merging**